### PR TITLE
fix(pkg/database): avoid silent returns when the scan limit is reached

### LINKED ISSUE #1202: [Bug] Error-prone scan usage if no limit is specified
It could lead to silent errors that pop up only once the default limit is reached.

### PR DESCRIPTION
**Pull request by jeroiraz**
Fixes #1202

Signed-off-by: Jeronimo Irazabal <jeronimo.irazabal@gmail.com>

### CODE REVIEW

File: pkg/database/database.go [hunk 1]
```diff
@@ -1224,7 +1225,7 @@ func (d *db) TxScan(req *schema.TxScanRequest) (*schema.TxList, error) {
 	}
 
 	if req.Limit > MaxKeyScanLimit {
-		return nil, ErrMaxKeyScanLimitExceeded
+		return nil, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitExceeded, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/database.go:1228]:
```suggestion
		return nil, fmt.Errorf("%w: given limit (%d) is larger than the maximum allowed one (%d)", ErrMaxKeyScanLimitExceeded, req.Limit, MaxKeyScanLimit)
```

File: pkg/database/database.go [hunk 2]
```diff
@@ -1261,6 +1262,10 @@ func (d *db) TxScan(req *schema.TxScanRequest) (*schema.TxList, error) {
 		}
 
 		txList.Txs = append(txList.Txs, sTx)
+
+		if l == MaxKeyScanLimit {
+			return txList, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/database.go:1267]:
```suggestion
			return txList, fmt.Errorf("%w: found more than %d entries (the maximum limit), please use the limit and initialTx arguments to paginate over large results", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```

File: pkg/database/database.go [hunk 3]
```diff
@@ -1273,7 +1278,7 @@ func (d *db) History(req *schema.HistoryRequest) (*schema.Entries, error) {
 	}
 
 	if req.Limit > MaxKeyScanLimit {
-		return nil, ErrMaxKeyScanLimitExceeded
+		return nil, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitExceeded, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/database.go:1281]:
```suggestion
		return nil, fmt.Errorf("%w: given limit (%d) is larger than the maximum allowed one (%d)", ErrMaxKeyScanLimitExceeded, req.Limit, MaxKeyScanLimit)
```

File: pkg/database/database.go [hunk 4]
```diff
@@ -1351,6 +1356,10 @@ func (d *db) History(req *schema.HistoryRequest) (*schema.Entries, error) {
 		}
 	}
 
+	if limit == MaxKeyScanLimit && hCount >= MaxKeyScanLimit {
+		return list, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/database.go:1360]:
```suggestion
			return list, fmt.Errorf("%w: found more than %d entries (the maximum limit), please use the limit and offset arguments to paginate over large results", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```

File: pkg/database/scan.go [hunk 5]
```diff
@@ -32,7 +34,7 @@ func (d *db) Scan(req *schema.ScanRequest) (*schema.Entries, error) {
 	}
 
 	if req.Limit > MaxKeyScanLimit {
-		return nil, ErrMaxKeyScanLimitExceeded
+		return nil, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitExceeded, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/scan.go:37]:
```suggestion
		return nil, fmt.Errorf("%w: given limit (%d) is larger than the maximum allowed one (%d)", ErrMaxKeyScanLimitExceeded, req.Limit, MaxKeyScanLimit)
```

File: pkg/database/scan.go [hunk 6]
```diff
@@ -100,13 +101,12 @@ func (d *db) Scan(req *schema.ScanRequest) (*schema.Entries, error) {
 			return nil, err
 		}
 
-		entries = append(entries, e)
-		if i++; i == limit {
-			break
+		entries.Entries = append(entries.Entries, e)
+
+		if l == MaxKeyScanLimit {
+			return entries, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/scan.go:107]:
```suggestion
			return entries, fmt.Errorf("%w: found more than %d entries (the maximum limit), please use the limit and seekKey arguments to paginate over large results", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```

File: pkg/database/sorted_set.go [hunk 7]
```diff
@@ -99,10 +100,10 @@ func (d *db) ZScan(req *schema.ZScanRequest) (*schema.ZEntries, error) {
 	}
 
 	if req.Limit > MaxKeyScanLimit {
-		return nil, ErrMaxKeyScanLimitExceeded
+		return nil, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitExceeded, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/sorted_set.go:103]:
```suggestion
		return nil, fmt.Errorf("%w: given limit (%d) is larger than the maximum allowed one (%d)", ErrMaxKeyScanLimitExceeded, req.Limit, MaxKeyScanLimit)
```

File: pkg/database/sorted_set.go [hunk 8]
```diff
@@ -233,17 +233,14 @@ func (d *db) ZScan(req *schema.ZScanRequest) (*schema.ZEntries, error) {
...
-		if i++; i == limit {
-			break
-		}
-	}
+		entries.Entries = append(entries.Entries, zentry)
 
-	list := &schema.ZEntries{
-		Entries: entries,
+		if l == MaxKeyScanLimit {
+			return entries, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/sorted_set.go:239]:
```suggestion
			return entries, fmt.Errorf("%w: found more than %d entries (the maximum limit), please use the limit, seekKey, seekScore and seekAtTx arguments to paginate over large results", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```

File: pkg/database/sql.go [hunk 9]
```diff
@@ -441,6 +437,10 @@ func (d *db) SQLQueryPrepared(stmt sql.DataSource, namedParams []*schema.NamedPa
 		}
 
 		res.Rows = append(res.Rows, rrow)
+
+		if l == MaxKeyScanLimit {
+			return res, fmt.Errorf("%w: %d is the limit", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```
Comment (byo) [inline pkg/database/sql.go:442]:
```suggestion
			return res, fmt.Errorf("%w: found more than %d rows (the maximum limit), please constrain the query with LIMIT clause", ErrMaxKeyScanLimitReached, MaxKeyScanLimit)
```

**Review by byo — COMMENTED** [review]
Added some suggestions to have better error messages (so that those can easily be understood by the user).

Also please add test cases where the limit is not set and the maximum limit is reached.

**Review by byo — APPROVED** [review]
